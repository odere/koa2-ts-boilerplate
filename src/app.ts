import cors = require('@koa/cors');
import koaBody = require('koa-body');
import koaBunyanLogger = require('koa-bunyan-logger');
import serve = require('koa-static');
import Koa from 'koa';
import { Server } from 'http';

import errorHadler from './middlewares/error-hendler';
import initData from './middlewares/init-data';
import { routes } from './routes';
import logger from './logger';

export interface AppOptions {
  env?: string;
  keys?: string[];
  proxy?: boolean;
  subdomainOffset?: number;
  proxyIpHeader?: string;
  maxIpsCount?: number;
}

export default class App extends Koa {
  servers: Server[];

  proxy: boolean;

  constructor(params?: AppOptions) {
    super(params);

    // Trust proxy
    this.proxy = true;

    this.servers = [];

    this.#configureMiddlewares();
    this.#configureLogger();
    this.#configureRoutes();
  }

  #configureMiddlewares(): void {
    this.use(koaBody());
    this.use(cors());
    this.use(errorHadler());
    this.use(initData());
    this.use(serve('public'));
  }

  #configureRoutes(): void {
    this.use(routes);
  }

  #configureLogger(): void {
    if (process.env.NODE_ENV !== 'test') {
      this.use(koaBunyanLogger(logger));
      this.use(koaBunyanLogger.requestLogger());
      this.use(koaBunyanLogger.timeContext());
    }
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  listen(...args): Server {
    const server = super.listen(...args);

    this.servers.push(server);

    return server;
  }

  terminate(): void {
    this.servers.forEach((server: Server) => {
      server.close();
    });
  }
}
