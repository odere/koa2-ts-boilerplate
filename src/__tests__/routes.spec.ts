import request from 'supertest';
import app from '../index';

describe('Routes', () => {
  afterAll(() => {
    app.terminate();
  });

  test('/', async () => {
    const response = await request(app.callback()).get('/');

    expect(response.status).toBe(401);
  });

  test('/health', async () => {
    const response = await request(app.callback()).get('/health');

    expect(response.status).toBe(200);
    expect(response.text).toMatchSnapshot();
  });
});
