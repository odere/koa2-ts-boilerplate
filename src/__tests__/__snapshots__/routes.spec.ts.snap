// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes /health 1`] = `"OK"`;
